### PR TITLE
feat: adds playbook and role to download and setup aws_logs_agent

### PR DESCRIPTION
**Pull request by DubeySandeep**
## Description

This PR adds a playbook for cloudwatch aws logs agent setup which will be further used for logging server logs to cloudwatch.

**Ticket link:** https://tasks.opencraft.com/browse/BB-6085

## Testing instructions
Prereq: All the aws resources are created for storing server logs on cloudwatch ([steps](https://github.com/open-craft/terraform-scripts/pull/40))
1. Add similar variable in any config file
```
CLOUDWATCH_LOG_FILES:
  - filename: /edx/var/log/lms/edx.log
    group_name: uq-staging-elasticsearch-cloudwatch-log-group
    stream_name: uq-staging-elasticsearch-cloudwatch-log-stream
  - filename: /edx/var/log/cms/edx.log
    group_name: uq-staging-elasticsearch-cloudwatch-log-group
    stream_name: uq-staging-elasticsearch-cloudwatch-log-stream

AWS_EC2_INSTANCE_REGION: ap-southeast-2
```
2. Run the following command to test the playbook
`ansible-playbook --extra-vars=@<filepath-to-config-variables> -i <inventory-path> -l <hosts-name> playbooks/cloudwatch_server_logs.yml `

### CODE REVIEW

File: playbooks/roles/aws_logs_agent/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    - install:base
+
+- name: Copy aws_logs_agent.conf template
+  template:
+    src: "aws_logs_agent.conf.j2"
+    dest: "/etc/aws_logs_agent.conf"
+    mode: "0664"
+
+- name: Run AWS Logs Agent script
+  shell: "python2.7 /tmp/aws-logs-agent.py -n --region {{ AWS_EC2_INSTANCE_REGION }} -c /etc/aws_logs_agent.conf"
```
Comment (0x29a) [inline playbooks/roles/aws_logs_agent/tasks/main.yml]: Python 2 has been removed from Ubuntu 20.04. Shouldn't we use Python 3 here?

Also, we can use the built-in [script module](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/script_module.html).

File: playbooks/roles/aws_logs_agent/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+---
+
+- name: Download the AWS Logs Agent python script
+  get_url:
+    url: "{{ AWS_LOGS_AGENT_SCRIPT_URL }}"
+    dest: /tmp/aws-logs-agent.py
+    mode: "1777"
```
Comment (0x29a) [inline playbooks/roles/aws_logs_agent/tasks/main.yml]: I'm not sure what this script requires, but isn't this too permissive?

File: playbooks/cloudwatch_server_logs.yml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+---
+
+- name: Set up aws_logs_agent to send server logs to cloudwatch
+  hosts: all
+  become: true
```
Comment (0x29a) [inline playbooks/cloudwatch_server_logs.yml]: Is it possible to use the `become` statement only for tasks that require admin privileges, inside the role? For example, `get_url` might not require them.

File: playbooks/roles/aws_logs_agent/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  tags:
+    - install
+    - install:base
```
Comment (0x29a) [inline playbooks/roles/aws_logs_agent/tasks/main.yml]: Are these tags are correct?

**Review by 0x29a — CHANGES_REQUESTED** [review]

File: playbooks/roles/aws_logs_agent/tasks/main.yml (hunk 1 above)
Comment (DubeySandeep) [reply playbooks/roles/aws_logs_agent/tasks/main.yml]: This script only works for `v2.6 < python > v3.6`, I found the servers either have v2.7 or v3.8, I'll check if I can download python v3.5 and run this script.

File: playbooks/cloudwatch_server_logs.yml (hunk 3 above)
Comment (DubeySandeep) [reply playbooks/cloudwatch_server_logs.yml]: Thanks for raising it, I had followed existing pattern but I agree that we should only allow sudo access to required steps. I'll look into it!

AFAIK, we only need sudo access to run the aws_logs_agent.py script.

File: playbooks/roles/aws_logs_agent/tasks/main.yml (hunk 2 above)
Comment (DubeySandeep) [reply playbooks/roles/aws_logs_agent/tasks/main.yml]: Changed it to "0664"

File: playbooks/roles/aws_logs_agent/tasks/main.yml (hunk 4 above)
Comment (DubeySandeep) [reply playbooks/roles/aws_logs_agent/tasks/main.yml]: Changed it to "install:code"

File: playbooks/roles/aws_logs_agent/tasks/main.yml (hunk 1 above)
Comment (DubeySandeep) [reply playbooks/roles/aws_logs_agent/tasks/main.yml]: steps to download python3.5
>Also, we can use the built-in script module.

I had tried script module, but it wasn't working correctly with the sudo rights and was constantly raising permission issues, so I tried shell and it worked fine (followed some existing roles structure.)

File: playbooks/cloudwatch_server_logs.yml (hunk 3 above)
Comment (DubeySandeep) [reply playbooks/cloudwatch_server_logs.yml]: Done!

**Comment by DubeySandeep** [comment]
@0x29a  I've addressed your review comments PTAL!

File: playbooks/roles/aws_logs_agent/tasks/main.yml (hunk 1 above)
Comment (0x29a) [reply playbooks/roles/aws_logs_agent/tasks/main.yml]: @DubeySandeep, I'm sorry for causing the back-and-forth, but can we avoid installing an additional interpreter for this? For example, I on [this page](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/download-cloudwatch-agent-commandline.html) I see URLs to download `.deb` packages. Could you try one of them? Please remember to use `ap-southeast-2` AWS region.

File: playbooks/roles/aws_logs_agent/tasks/main.yml (hunk 4 above)
Comment (0x29a) [reply playbooks/roles/aws_logs_agent/tasks/main.yml]: @DubeySandeep, sorry, I should have been more clear here. These tags are needed to run some specific set of Ansible tasks. What is the goal of tagging this task alone?

File: playbooks/roles/aws_logs_agent/tasks/main.yml (hunk 1 above)
Comment (DubeySandeep) [reply playbooks/roles/aws_logs_agent/tasks/main.yml]: > can we avoid installing an additional interpreter for this?

@0x29a we already have some steps which requires py3.5 ([ref](https://github.com/openedx/configuration/blob/4a98e09fa8a388cdcf627299e7c9db4bbddc3748/playbooks/roles/common_vars/defaults/main.yml#L141))

I'm following the official guide to installing and setup awslogs-agent ([ref](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/QuickStartEC2Instance.html)) and the script does all the required installation/setup with one command.

I'm looking into how I can use the .deb package to configure the agent to publish logs, I'll update the thread once I found a way to do it!

Comment (DubeySandeep) [reply playbooks/roles/aws_logs_agent/tasks/main.yml]: I've found a way to use the Debian package, will update the changes and get it ready for review!

Comment (DubeySandeep) [reply playbooks/roles/aws_logs_agent/tasks/main.yml]: @0x29a I've updated the role to use .deb package

File: playbooks/roles/aws_logs_agent/tasks/main.yml (hunk 4 above)
Comment (DubeySandeep) [reply playbooks/roles/aws_logs_agent/tasks/main.yml]: I've removed it as it's not needed!

**Comment by DubeySandeep** [comment]
@0x29a I've addressed your review comments, PTAL!

**Comment by DubeySandeep** [comment]
Suppressed by https://github.com/openedx/configuration/pull/6767

**Pull request closed without merging**